hack/test-go.sh

```diff
@@ -1,124 +1,177 @@
 
-  exit 0
+    if [[ -n "${test_kube}" ]]; then
+        test_packages="$(echo "${test_packages}"; echo "${optional_kubernetes_packages}")"
+    fi
 fi
 
-export OPENSHIFT_ON_PANIC=crash
+if [[ -n "${print_packages}" ]]; then
+    echo "The following packages will be tested by $0:"
+    for package in "${test_packages}"; do
+        echo "${package}"
+    done
+fi
 
-if [[ -n "${KUBE_COVER}" && -n "${OUTPUT_COVERAGE}" ]]; then
-  # Iterate over packages to run coverage
-  test_packages=( $test_packages )
-  for test_package in "${test_packages[@]}"
-  do
-    mkdir -p "$OUTPUT_COVERAGE/$test_package"
-    KUBE_COVER_PROFILE="-coverprofile=$OUTPUT_COVERAGE/$test_package/profile.out"
+# Run 'go test' with the accumulated arguments and packages:
+# we can run 'go test' alone, or also generate coverage reports, or generate jUnit XML reports
+if [[ -n "${junit_report}" ]]; then
+    if [[ -z "$(which junitreport)" ]]; then
+        # we need to build the 'junitreport' tool
+        "${openshift_root_dir}/hack/build-go.sh tools/junitreport"
+    fi
 
-    go test $KUBE_RACE -timeout $KUBE_TIMEOUT $KUBE_COVER "$KUBE_COVER_PROFILE" "$test_package" "${@:2}"
-  done
+    # we need to generate jUnit XML
+    test_output_file="${TEMPDIR}/test-go.out"
+    junit_report_file="${TEMPDIR}/report.xml"
+    go test ${gotest_flags} ${test_packages} > ${test_output_file} 2>&1
 
-  echo 'mode: atomic' > ${OUTPUT_COVERAGE}/profiles.out
-  find $OUTPUT_COVERAGE -name profile.out | xargs sed '/^mode: atomic$/d' >> ${OUTPUT_COVERAGE}/profiles.out
-  go tool cover "-html=${OUTPUT_COVERAGE}/profiles.out" -o "${OUTPUT_COVERAGE}/coverage.html"
+    suite_roots="github.com/openshift/origin,github.com/openshift/origin/Godeps/_workspace/src/k8s.io/kubernetes"
+    cat "${test_output_file}" | junitreport --type gotest --suites nested --roots "${suite_roots}" > "${junit_report_file}"
+    cat "${junit_report_file}" | junitreport summarize
 
-  rm -rf $OUTPUT_COVERAGE/$OS_GO_PACKAGE
+elif [[ -n "${coverage_output_dir}" ]]; then
+    # we need to generate coverage reports
+    for test_package in "${test_packages}"; do
+        mkdir -p "${coverage_output_dir}/${test_package}"
+        gotest_flags="${gotest_flags} -coverprofile=${coverage_output_dir}/${test_package}/profile.out"
+
+        go test ${gotest_flags} ${test_package}
+
+        # assemble all profiles and generate a coverage report
+        echo 'mode: atomic' > "${coverage_output_dir}/profiles.out"
+        find "${coverage_output_dir}" -name profile.out | xargs sed '/^mode: atomic$/d' >> "${coverage_output_dir}/profiles.out"
+
+        go tool cover "-html=${coverage_output_dir}/profiles.out" -o "${coverage_output_dir}/coverage.html"
+
+        rm -rf "${coverage_output_dir}/${OS_GO_PACKAGE}"
+    done
 else
-  nice go test $KUBE_RACE -timeout $KUBE_TIMEOUT $KUBE_COVER "${@:2}" $test_packages
+    # we need to generate neither jUnit XML nor coverage reports
+    go test ${gotest_flags} ${test_packages}
```

Comment:
`go test` doesn't expose any flags with spaces and neither does it elect any flags to be set to values containing spaces (aside from paths, which the user will need to handle themselves). The lack of double quotes here is intentional. If we double quoted any of these variables we would need to `eval` the expression to parse it correctly.